htmd/qm/base.py

```diff
@@ -0,0 +1,202 @@
+        self._arg('queue', ':class:`SimQueue <htmd.queues.simqueue.SimQueue>` object',
+                  'Queue object used to run simulations',
+                  default=LocalCPUQueue())
+        self._arg('directory', 'str', 'Working directory',
+                  default='.', validator=val.String())
+
+    @property
+    @abstractmethod
+    def _command(self):
+        pass
+
+    @abstractmethod
+    def _completed(self, directory):
+        pass
+
+    @abstractmethod
+    def _writeInput(self, directory, iframe):
+        pass
+
+    @abstractmethod
+    def _readOutput(self, directory):
+        pass
+
+    def _setup(self):
+
+        # Set up the molecule
+        # TODO remove molecule coping!
+        self._molecule = self.molecule.copy()
+        self._nframes = self._molecule.coords.shape[2]
+        self._natoms = self._molecule.coords.shape[0]
+        self._charge = self._molecule.netcharge
+
+        # Set up ESP points
+        if self.esp_points is not None:
+            # TODO move to a validator
+            if self.esp_points.shape[1] != 3:
+                raise ValueError("ESP point array must be npoints x 3")
+            if self._molecule.coords.shape[2] != 1:
+                raise ValueError("Can only specift ESP point array with a single frame of coords")
+
+        # Set up restrained dihedrals
+        self._restrained_dihedrals = None
+        if self.restrained_dihedrals is not None:
+            self._restrained_dihedrals = self.restrained_dihedrals + 1  # Convert to 1-based indices
+
+        # TODO extract from SimQueue object
+        try:
+            self._ncpus = int(os.getenv('NCPUS'))
+        except TypeError:
+            self._ncpus = os.cpu_count()
+
+        # TODO extract from SimQueue object
+        self._mem = 2
+
+        # Create directories and write inputs
+        self._directories = []
+        for iframe in range(self._nframes):
+
+            # Create a directory
+            directory = os.path.join(self.directory, '%05d' % iframe)
+            os.makedirs(directory, exist_ok=True)
+            self._directories.append(directory)
+
+            if not self._completed(directory):
+
+                # Write input files
+                self._writeInput(directory, iframe)
+
+                # Write a point file for ESP
+                if self.esp_points is not None:
+                    np.savetxt(os.path.join(directory, 'grid.dat'), self.esp_points, fmt='%f')
+
+                # Write a run script
+                script = os.path.join(directory, 'run.sh')
+                with open(script, 'w') as f:
+                    f.write('#!/bin/sh\n\n%s\n' % self._command)
+                os.chmod(script, 0o700)
+
+    def _submit(self):
+
+        for directory in self._directories:
+            if not self._completed(directory):
+                self.queue.submit(directory)
+
+    def _retrieve(self):
+
+        self.queue.wait()
```

Comment:
The `queue.wait(sentinel=True)` is important, specially for non-localqueues. The idea behind using it is described in the wait method of LsfQueue ("If False, it relies on the queueing system reporting to determine the number of running jobs. If True, it relies on the filesystem, in particular on the existence of a sentinel file for job completion."). The reason being, sometimes there's an inconsistency between what the queue reports and the files already being on the system. This means that an entire job can die on when trying to read the output file.
I will add the changes to the LocalQueue (to have a correct method) and also add it to the Base Class SimQueue in the master branch, and then you can pull.

Reply:
Pending...

Reply:
@j3mdamas 

It seems the sentinel stuff is not working with 'LocalCPUQueue`:

```python
Traceback (most recent call last):
  File "/home/raimis/opt/miniconda3/bin/parameterize", line 11, in <module>
    load_entry_point('htmd', 'console_scripts', 'parameterize')()
  File "/home/raimis/prj/htmd.git/htmd/parameterization/cli.py", line 177, in main_parameterize
    mol.minimize()
  File "/home/raimis/prj/htmd.git/htmd/parameterization/ffmolecule.py", line 164, in minimize
    results = self.qm.run()
  File "/home/raimis/prj/htmd.git/htmd/qm/base.py", line 189, in run
    return self._retrieve()
  File "/home/raimis/prj/htmd.git/htmd/qm/base.py", line 166, in _retrieve
    self.queue.wait(sentinel=True)
  File "/home/raimis/prj/htmd.git/htmd/queues/simqueue.py", line 53, in wait
    while (self.inprogress() if not sentinel else self.notcompleted()) != 0:
  File "/home/raimis/prj/htmd.git/htmd/queues/localqueue.py", line 197, in notcompleted
    raise RuntimeError('This method relies on running synchronously.')
RuntimeError: This method relies on running synchronously.
```